common/web/lm-worker/src/main/correction/context-tracker.ts

```diff
@@ -391,100 +413,85 @@ export class ContextTracker extends CircularArray<TrackedContextState> {
-          *
-          * Less-common, but noteworthy:  some wordbreaks may occur without whitespace.  Example:
-          * `"o` => ['"', 'o'].  Make sure to double-check against `tokenizedContext`!
-          */
-        let pushedToken = new TrackedContextToken();
-        pushedToken.raw = tokenizedTail;
-
-        if(isWhitespace || !primaryInput) {
-          state.pushWhitespaceToTail(transformDistribution ?? []);
-          // Continuing the earlier assumption, that 'pure-whitespace Transform' does not emit any initial characters
-          // for the new word (token), so the input keystrokes do not correspond to the new text token.
-          pushedToken.transformDistributions = [];
-        } else {
-          state.pushWhitespaceToTail();
-          // Assumption: Since we only allow one-transform-at-a-time changes between states, we shouldn't be missing
-          // any metadata used to construct the new context state token.
-          pushedToken.transformDistributions = transformDistribution ? [transformDistribution] : [];
-        }
+    // Now to update the end of the context window.
+    for(let i = lastMatch+1; i < editPath.length; i++) {
+      const incomingToken = tokenizedContext[i - poppedTokenCount]
+      switch(editPath[i]) {
+        case 'substitute':
+          if(i == editPath.length - 1) {
+            state = new TrackedContextState(state);
+          }
 
-        state.pushTail(pushedToken);
-      } else {
-        // We're editing the final context token.
-        // TODO:  Assumption:  we didn't 'miss' any inputs somehow.
-        //        As is, may be prone to fragility should the lm-layer's tracked context 'desync' from its host's.
-        maintainLastToken();
-      }
-      // There is only one word in the context.
-    } else {
-      // TODO:  Assumption:  we didn't 'miss' any inputs somehow.
-      //        As is, may be prone to fragility should the lm-layer's tracked context 'desync' from its host's.
-
-      if(editPath[tailIndex] == 'insert') {
-        // Construct appropriate initial token.
-        let token = new TrackedContextToken();
-        token.raw = tokenizedContext[0];
-        token.transformDistributions = [transformDistribution];
-        state.pushTail(token);
-      } else {
-        // Edit the lone context token.
-        maintainLastToken();
+          if(isBackspace) {
+            state.tokens[i - poppedTokenCount].updateWithBackspace(incomingToken.text, primaryInput.id);
+          } else {
+            state.updateToken(state.tokens[i - poppedTokenCount], transformDistribution, incomingToken.text);
+          }
+
+          if(state != matchState) {
+            if(isBackspace) {
+              matchState.tokens[i].updateWithBackspace(incomingToken.text, primaryInput.id);
+            } else {
+              matchState.updateToken(state.tokens[i], transformDistribution, incomingToken.text);
```

Comment:
```suggestion
              matchState.updateToken(matchState.tokens[i], transformDistribution, incomingToken.text);
```

Fixed in #11997 already, though.